sdk offline caching considerations (#509)
getsentry/develop: src/docs/sdk/features.mdx

## Buffer to Disk

This feature is also known as 'Offline Caching'.

> Write events to disk before attempting to send, so that they can be retried in the event of a temporary network failure. Needs to implement a cap on the number of stored events. This is mostly useful on mobile and desktop(e.g: laptop) apps, where stable connectivity is often not available.

Reviewer: Would it be useful to link to some implementations where this is working? Like to cocao, .NET, electron?

Author: Added 4 examples below (linked to the file on the main branch instead of specific lines and a commit sha, so technically could go out of date if the file move or main branch changes but better to point to "latest" I believe

> #### Deailing with failures

Reviewer: ```suggestion
#### Dealing with failures
```

It's important to note that retry is only considered in the event of a network failure. For example:

* Connection timeout
* DSN resolution failure
* Connection reset by peer

Other failures, like those caused by processing the file in the SDK itself, the payload should be discarded since those are likely to end up on an endless retry.

> If the event reached Sentry and a HTTP response status code was received, even if in the event of a `500`, the event should be discarded.

Reviewer: ```suggestion
If the event reached Sentry and a HTTP response status code was received, even in the event of a `500` response, the event should be discarded.
```

#### Additional capabilities

Consider having the SDK retry sending events once the device is back online, when such notification exists in the platform.

> Once the device is back online, the SDK is likely going to empty its disk queue in a quick burst of requests. This can trigger different abuse filters in Sentry. To account for that, it's considered to add a small latency between cached event captures. A recommended value is 100 milliseconds.

Author: ```suggestion
Once the device is back online, the SDK is likely going to empty its disk queue in a quick burst of requests. This can trigger different abuse filters in Sentry. To account for that, it's considered to add a small delay between cached event captures. A recommended value is 100 milliseconds.
```

> If the SDK is being [rate-limited](/sdk/rate-limiting/), which causes the SDK to drop any event that reaches its HTTP transport, cosider stop consuming the disk cache until the backoff timeout is reached or the app restarts.

Author: ```suggestion
If the SDK is being [rate-limited](/sdk/rate-limiting/), which causes the SDK to drop any event that reaches its HTTP transport, cosider stop consuming the disk cache until the `Retry-After` timeout is reached or the app restarts.
```

> If the SDK is being [rate-limited](/sdk/rate-limiting/), which causes the SDK to drop any event that reaches its HTTP transport, cosider stop consuming the disk cache until the `Retry-After` timeout is reached or the app restarts.

Author: ```suggestion
If the SDK is being [rate-limited](/sdk/rate-limiting/), which causes the SDK to drop any event that reaches its HTTP transport, cosider stop consuming the disk cache until the `Retry-After` timeout is reached or the app restarts.

#### Example implementations

- [C#](https://github.com/getsentry/sentry-dotnet/blob/main/src/Sentry/Internal/Http/CachingTransport.cs)
- [Java](https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/cache/EnvelopeCache.java)
- [Objective-C](https://github.com/getsentry/sentry-cocoa/blob/master/Sources/Sentry/SentryHttpTransport.m)
- [TypeScript](https://github.com/getsentry/sentry-electron/blob/master/src/main/transports/electron-offline-net.ts)
```

## Buffer to Disk

This feature is also known as 'Offline Caching'.

Write events to disk before attempting to send, so that they can be retried in the event of a temporary network failure. Needs to implement a cap on the number of stored events. This is mostly useful on mobile and desktop(e.g: laptop) apps, where stable connectivity is often not available.

#### Dealing with failures

It's important to note that retry is only considered in the event of a network failure. For example:

* Connection timeout
* DSN resolution failure
* Connection reset by peer

Other failures, like those caused by processing the file in the SDK itself, the payload should be discarded since those are likely to end up on an endless retry.
If the event reached Sentry and a HTTP response status code was received, even in the event of a `500` response, the event should be discarded.

#### Additional capabilities

Consider having the SDK retry sending events once the device is back online, when such notification exists in the platform.

> Once the device is back online, the SDK is likely going to empty its disk queue in a quick burst of requests. This can trigger different abuse filters in Sentry. To account for that, it's considered to add a small delay between cached event captures. A recommended value is 100 milliseconds.

Author: It isn't in any. We have issues and some notion items to plan this in the future

If the SDK is being [rate-limited](/sdk/rate-limiting/), which causes the SDK to drop any event that reaches its HTTP transport, cosider stop consuming the disk cache until the `Retry-After` timeout is reached or the app restarts.

#### Example implementations

- [C#](https://github.com/getsentry/sentry-dotnet/blob/main/src/Sentry/Internal/Http/CachingTransport.cs)

> - [Java](https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/cache/EnvelopeCache.java)

Author: It's just a pointer to folks, didn't want to list all classes needed but something folks can find references from
